overlord/devicestate/systems.go

```diff
@@ -176,3 +185,231 @@ func seededSystemFromModeenv() (*seededSystem, error) {
 	}
 	return seededSys, nil
 }
+
+// createSystemForModelFromValidatedSnaps creates a new recovery system for the
+// specified model with the specified label using the snaps in the database and
+// the getInfo function. getInfo is expected to return for a given snap name
+// a snap.Info for that snap and whether the snap is present is present in the
+// recovery system. The second bit is relevant for optional snaps - required and
+// essential snaps from the model must always be present in the recovery system,
+// and it is an error for an essential or required snap from the model to be
+// returned as not present in the recovery system from getInfo.
+// The function returns the directory of the new recovery system as well as the
+// set of absolute file paths to the new snap files that were written for the
+// recovery system - some snaps may be in the recovery system directory while
+// others may be in the common snaps directory shared between multiple recovery
+// systems on ubuntu-seed.
+func createSystemForModelFromValidatedSnaps(model *asserts.Model, label string, db asserts.RODatabase, getInfo func(name string) (*snap.Info, bool, error)) (newFiles []string, dir string, err error) {
+	if model.Grade() == asserts.ModelGradeUnset {
+		return nil, "", fmt.Errorf("cannot create a system for non UC20 model")
+	}
+
+	logger.Noticef("creating recovery system with label %q for %q", label, model.Model())
+
+	// TODO: should that path provided by boot package instead?
+	recoverySystemDirInRootDir := filepath.Join("/systems", label)
+	assertedSnapsDir := filepath.Join(boot.InitramfsUbuntuSeedDir, "snaps")
+	recoverySystemDir := filepath.Join(boot.InitramfsUbuntuSeedDir, recoverySystemDirInRootDir)
+
+	wOpts := &seedwriter.Options{
+		// RW mount of ubuntu-seed
+		SeedDir: boot.InitramfsUbuntuSeedDir,
+		Label:   label,
+	}
+	w, err := seedwriter.New(model, wOpts)
+	if err != nil {
+		return nil, "", err
+	}
+
+	optsSnaps := make([]*seedwriter.OptionsSnap, 0, len(model.RequiredWithEssentialSnaps()))
+	// collect all snaps that are present
+	modelSnaps := make(map[string]*snap.Info)
+
+	getModelSnap := func(name string, essential bool, nonEssentialPresence string) error {
+		kind := "essential"
+		if !essential {
+			kind = "non-essential"
+			if nonEssentialPresence != "" {
+				kind = fmt.Sprintf("non-essential but %q",
```

Comment:
(nitpick^3) the linebreak here for the parameter looks not really needed, this is not overly long even with `nonEssentialPresence` in the same line

Reply:
tweaked